Gravatar/MruImageCache.cs

```diff
@@ -0,0 +1,119 @@
+{
+    /// <summary>
+    /// Caches most-recently-used images.
+    /// </summary>
+    /// <remarks>
+    /// Decorates an inner cache, delegating to it as needed.
+    /// <para />
+    /// If an image is available in memory, the inner cache can be bypassed.
+    /// </remarks>
+    public sealed class MruImageCache : IImageCache
+    {
+        private const int CleanAtSize = 30;
+        private const int CleanToSize = 25;
+
+        private readonly Dictionary<string, Entry> _entryByFileName = new Dictionary<string, Entry>();
+        private readonly IImageCache _inner;
+
+        public MruImageCache(IImageCache inner)
+        {
+            _inner = inner;
+        }
+
+        event EventHandler IImageCache.Invalidated
+        {
+            add => _inner.Invalidated += value;
+            remove => _inner.Invalidated -= value;
+        }
+
+        void IImageCache.AddImage(string imageFileName, Image image)
+        {
+            _inner.AddImage(imageFileName, image);
+            UpdateEntry(imageFileName, image);
+        }
+
+        Task IImageCache.ClearAsync()
+        {
+            _entryByFileName.Clear();
+            return _inner.ClearAsync();
+        }
+
+        Task IImageCache.DeleteImageAsync(string imageFileName)
+        {
+            _entryByFileName.Remove(imageFileName);
+            return _inner.DeleteImageAsync(imageFileName);
+        }
+
+        Image IImageCache.GetImage(string imageFileName)
+        {
+            if (_entryByFileName.TryGetValue(imageFileName, out var entry))
+            {
+                entry.LastAccesedAt = DateTime.UtcNow;
+                return entry.Image;
+            }
+
+            return _inner.GetImage(imageFileName);
+        }
+
+        async Task<Image> IImageCache.GetImageAsync(string imageFileName)
+        {
+            if (_entryByFileName.TryGetValue(imageFileName, out var entry))
+            {
+                entry.LastAccesedAt = DateTime.UtcNow;
+                return entry.Image;
+            }
+
+            var image = await _inner.GetImageAsync(imageFileName);
+
+            UpdateEntry(imageFileName, image);
+
+            return image;
+        }
+
+        private void UpdateEntry(string imageFileName, Image image)
+        {
+            _entryByFileName[imageFileName] = new Entry(imageFileName, image);
+
+            if (_entryByFileName.Count > CleanAtSize)
+            {
+                RemoveOldEntries();
+            }
+
+            void RemoveOldEntries()
+            {
+                var sortedEntries = new SortedList<DateTime, Entry>(_entryByFileName.Count);
+
+                foreach (var entry in _entryByFileName.Values)
+                {
+                    sortedEntries.Add(entry.LastAccesedAt, entry);
+                }
+
+                foreach (var entry in sortedEntries.Values.Take(_entryByFileName.Count - CleanToSize))
+                {
+                    _entryByFileName.Remove(entry.FileName);
+                }
```

Comment:
Algorithmically it might be. It involves allocating more memory, and is O(NlogN) rather than O(1) in the worst case. `Values` is ordered, so it might just be O(N). Either way, it's more memory and more CPU to do it that way.

Reply:
My last comment was shooting from the hip based on the 4 lines  of code GitHub lets you see in the browser. I should have pulled in more context. I think what I wrote is largely garbage, sorry!

Reply:
In the current implementation, `sortedEntries` needs to hold all entries so that we can pull out the oldest ones (they are sorted to the start of the collection). `_entryByFileName` does give some ordering guarantees, but as the `LastAccessedAt` can change, the collection's ordering doesn't necessarily match the order of entries by that property. So in this case I can't see a way to simplify the code here.

I'll capture this in comments in the code itself.